### PR TITLE
[IMP] stock: Add context key to avoid send sms

### PR DESCRIPTION
**Pull request by jbm-odoo**
In stock, before sending an sms, we check if there aren't the key
'skip_sms' in the context. It is useful, for example, in FSM.
application.

TaskID: 2081191

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock_account/models/product.py [hunk 1]
```diff
@@ -605,6 +605,9 @@ def _compute_average_price(self, qty_invoiced, qty_to_invoice, stock_moves):
         """
         self.ensure_one()
 
+        if not qty_to_invoice:
+            return 0
```
Comment (sle-odoo) [inline addons/stock_account/models/product.py]: what?

File: addons/stock_sms/models/stock_picking.py [hunk 2]
```diff
@@ -19,6 +19,8 @@ def _pre_action_done_hook(self):
 
     def _check_warn_sms(self):
         warn_sms_pickings = self.browse()
+        if self.env.context.get('skip_sms'):
+            return warn_sms_pickings
```
Comment (sle-odoo) [inline addons/stock_sms/models/stock_picking.py]: if you add a skip_* context key, could you add at a consistent place?
if you look at stock picking _pre_action_done_hook
``` python
    def _pre_action_done_hook(self):
        if not self.env.context.get('skip_immediate'):
            pickings_to_immediate = self._check_immediate()
            if pickings_to_immediate:
                return pickings_to_immediate._action_generate_immediate_wizard(show_transfers=self._should_show_transfers())

        if not self.env.context.get('skip_backorder'):
            pickings_to_backorder = self._check_backorder()
            if pickings_to_backorder:
                return pickings_to_backorder._action_generate_backorder_wizard(show_transfers=self._should_show_transfers())
        return True
```
so it seems to me the right place would be in the same method which currently looks like it

``` python
    def _pre_action_done_hook(self):
        res = super()._pre_action_done_hook()
        if res is True:
            pickings_to_warn_sms = self._check_warn_sms()
            if pickings_to_warn_sms:
                return pickings_to_warn_sms._action_generate_warn_sms_wizard()
        return res
```

File: addons/stock_account/models/product.py (hunk 1 above)
Comment (jbm-odoo) [reply addons/stock_account/models/product.py]: Without this check, we can have a division by zero in line "return tmp_value / qty_to_invoice". (traceback: https://pastebin.com/W9k120Hy)

Comment (sle-odoo) [reply addons/stock_account/models/product.py]: according to your traceback, this call triggers the issue
``` python
line.purchase_price = line.product_id._compute_average_price(0, line.product_uom_qty, line.move_ids)
```
so qty_to_invoice is line.product_uom_qty and this field cannot be equals to zero
``` python
product_uom_qty = fields.Float(string='Quantity', digits='Product Unit of Measure', required=True, default=1.0)
```

how can i reproduce this traceback in master ?

Comment (jbm-odoo) [reply addons/stock_account/models/product.py]: Here is a video to reproduce this traceback in master https://drive.google.com/open?id=1nLt4sBGGaZuzjP3f8bAJogfTcpfvIZ5N

Comment (sle-odoo) [reply addons/stock_account/models/product.py]: it seems float required are useless.. so fine by me

**Comment by tivisse** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**